support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid.go

```diff
@@ -0,0 +1,180 @@
+limitations under the License.
+*/
+
+package webhook
+
+import (
+	"bytes"
+	"context"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"net/url"
+	"os"
+	"path"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"github.com/spiffe/go-spiffe/v2/spiffeid"
+	svidv1 "github.com/spiffe/spire-api-sdk/proto/spire/api/server/svid/v1"
+	spiretypes "github.com/spiffe/spire-api-sdk/proto/spire/api/types"
+)
+
+const (
+	certDirMode   = os.FileMode(0o600)
+	certsFileMode = os.FileMode(0o644)
+	keyFileMode   = os.FileMode(0o600)
+	certsFileName = "tls.crt"
+	keyFileName   = "tls.key"
+)
+
+type SVIDConfig struct {
+	Cluster            string
+	Log                logrus.FieldLogger
+	Namespace          string
+	S                  svidv1.SVIDClient
+	TrustDomain        string
+	WebhookCertDir     string
+	WebhookServiceName string
+}
+
+type SVID struct {
+	c         SVIDConfig
+	id        spiffeid.ID
+	beginsAt  time.Time
+	expiresAt time.Time
+}
+
+// NewSVID sets up the data structure and mints a new SVID for the webhok
+func NewSVID(ctx context.Context, config SVIDConfig) (*SVID, error) {
+	err := os.MkdirAll(config.WebhookCertDir, certDirMode)
+	if err != nil {
+		return nil, err
+	}
+
+	id, err := spiffeid.New(config.TrustDomain, "k8s-workload-registrar", config.Cluster, "webhook")
+	if err != nil {
+		return nil, fmt.Errorf("unable to generate Spiffe ID: %w", err)
+	}
+
+	return &SVID{
+		c:  config,
+		id: id,
+	}, nil
+}
+
+// MintSVID requests the SPIRE Server to mint a new SVID for the webhook
+func (e *SVID) MintSVID(ctx context.Context, key crypto.Signer) (err error) {
+	e.c.Log.Debug("Minting new SVID for webhook")
+	// Generate key if not passed in
+	if key == nil {
+		key, err = ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+		if err != nil {
+			return fmt.Errorf("unable to generate Key: %w", err)
+		}
+	}
+
+	// Generate Certificate Signing Request
+	csr, err := x509.CreateCertificateRequest(rand.Reader, &x509.CertificateRequest{
+		URIs: []*url.URL{e.id.URL()},
+		DNSNames: []string{
+			e.c.WebhookServiceName + "." + e.c.Namespace + ".svc",
+		},
+	}, key)
+	if err != nil {
+		return fmt.Errorf("unable to generate CSR: %w", err)
+	}
+
+	// Mint new SVID
+	resp, err := e.c.S.MintX509SVID(ctx, &svidv1.MintX509SVIDRequest{
+		Csr: csr,
+	})
+	if err != nil {
+		return fmt.Errorf("unable to make Mint SVID Request: %w", err)
+	}
+	e.beginsAt = time.Now()
+	e.expiresAt = time.Unix(resp.Svid.GetExpiresAt(), 0)
+
+	// Write SVID and key to disk
+	return e.dumpSVID(resp.Svid, key)
+}
+
+// SVIDRotator requests a new certificate when half of its lifetime is left
+func (e *SVID) SVIDRotator(ctx context.Context) error {
+	for {
+		lifetime := e.expiresAt.Sub(e.beginsAt) / 2
+		ttl := time.Until(e.expiresAt)
+		select {
+		case <-time.After(ttl - lifetime):
+			if err := e.MintSVID(ctx, nil); err != nil {
+				return fmt.Errorf("unable to mint SVID: %w", err)
```

Comment:
See my other comment. If we change this code to no longer cycle the registrar on failure, then we'll also need to implement a backoff strategy so we don't pummel the server with requests.

Reply:
Yes good point. How about using `retry.OnError()`? https://pkg.go.dev/k8s.io/client-go/util/retry#OnError

Reply:
That's probably ok? What do you plan on doing inside of the "retriable" callback?

Reply:
These lines of code: https://github.com/spiffe/spire/pull/2321/files#diff-ff4a359cdb6e273cd555b8aa198c0d14bed25a991da06708c8e7453c4642dc8bR104-R115

Reply:
That seems like what you'd put in `fn`. `retriable` classifies the given error as retriable or not. I was just curious if you were going to classify all errors as retriable or what.

Reply:
Is there an error we don't want to retry on?

Reply:
I don't think so. One thing we may consider doing is returning false if the current SVID has expired, which will cause 
OnError to return an error, which we can then use to tear down the registrar (since the webhook is now presumably non-functional).

Reply:
This change is in.